### ISSUE TITLE
Can the features of Query Cache support subqueries?

### ISSUE THREAD

**Issue by lzf575** [ClickHouse/ClickHouse#51656]
I see that the feature of Query Cache has been completed
I tried it, and it supports the result cache of the entire statement, but does not support the result cache of subquery?
I would like to ask if it（cache subquery）is not supported, or it is my usage problem.
  - the SQL statement `select a from (SubQueryA setting use_query_cache = 1)`, if the result of `SubQueryA` can be cached, then the execution performance of the statement `select b from (SubQueryA)` can also be improved?

If it is not supported, please tell me when it will be supported in the future？

-  the featue in changelog：![image](https://github.com/ClickHouse/ClickHouse/assets/44988793/260ac733-3144-47a4-a9fc-e559c674162c)

**Comment by alexey-milovidov** [comment]
Yes, this is doable and looks like a small addition to implement.

**Comment by vmihailenco** [comment]
This would help with incremental processing of large queries tremendously.

**Comment by rschu1ze** [comment]
This [talk](https://www.youtube.com/watch?v=4y656idl8ZU) by Firebolt gives some clues how to do it.

**Comment by volkandkaya** [comment]
This would reduce query times a lot for my use case where customers hit 10 endpoints every page refresh.

75% each query, going from 0.2 seconds to 0.05 seconds per query. For longer time frames/filters would be even larger.

**Comment by volkandkaya** [comment]
> This [talk](https://www.youtube.com/watch?v=4y656idl8ZU) by Firebolt gives some clues how to do it.

Where is the code for Clickhouse?

**Comment by rschu1ze** [comment]
You can send a PR that implements subquery caching. Since this topic is part of https://github.com/ClickHouse/ClickHouse/issues/71175 there is hope that a student will implement it.

**Comment by volkandkaya** [comment]
> You can send a PR that implements subquery caching. Since this topic is part of #71175 there is hope that a student will implement it.

Okay with the extra context the reply makes sense.

**Comment by rschu1ze** [comment]
Related: https://github.com/ClickHouse/ClickHouse/issues/93756

**Comment by alexey-milovidov** [comment]
Implemented in #99804
